### PR TITLE
Prepare integration tests on kubeflow

### PR DESCRIPTION
**Pull request by schustmi**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
- Pin the K3S image to a specific version to make sure the kubernetes/KFP versions match up
- Raise an exception when the call to `zenml example run` fails
- Add option to prevent stack setup when running an example
- Allow running the kubeflow orchestrator synchronously
- Detect whether a docker image is local and don't try to pull it in that case

### CODE REVIEW

File: tests/integration/test_examples_on_kubeflow.py [hunk 1]
```diff
@@ -0,0 +1,348 @@
+#  Copyright (c) ZenML GmbH 2021. All Rights Reserved.
```
Comment (AlexejPenner) [inline tests/integration/test_examples_on_kubeflow.py:1]:
```suggestion
#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
```

File: tests/integration/test_examples_on_kubeflow.py [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+    ) + [str(examples_dir / EXAMPLES_RUN_SCRIPT)]
+
+
+@pytest.mark.parametrize(
+    "example_configuration",
+    [pytest.param(example, id=example.name) for example in examples],
+)
+def test_run_example_on_kfp(
+    example_configuration: ExampleIntegrationTestConfiguration,
+    clean_kubeflow_repo,
```
Comment (AlexejPenner) [inline tests/integration/test_examples_on_kubeflow.py:324]: Could it make sense to combine the kubeflow integration tests with the local integration tests to avoid duplicating the test that have to be written for a new example? Then we could just run test_run_example_on_kfp or test_run_example_locally

**Review by AlexejPenner — CHANGES_REQUESTED** [review]

File: tests/integration/test_examples_on_kubeflow.py (hunk 2 above)
Comment (schustmi) [reply tests/integration/test_examples_on_kubeflow.py:324]: That was the initial goal, but if we only want to run the kubeflow tests e.g. on develop but the local integration test every time, we'll need to figure out a way to separate them. I have some ideas for that, but wanted to leave the existing stuff intact until both your and my PR are ready/merged

Comment (AlexejPenner) [reply tests/integration/test_examples_on_kubeflow.py:324]:
```bash
pytest tests/integration/test_examples_on_kubeflow.py::test_run_example_on_kfp
```
That one would only run the kubeflow examples. Similarly we could only run local examples

Comment (schustmi) [reply tests/integration/test_examples_on_kubeflow.py:324]: Ah okay I misunderstood the request I think, I thought merging them into a single test function (That would have the additional benefit of potentially only installing the requirements once into the separate virtual environment).If you think it doesn't lead to any complicated merge conflict with your branch(es), I can certainly merge them into a single file and use my example configurations to reduce the duplicated code

**Review by htahir1 — APPROVED** [review]
Very smart changes! LGTM!

File: src/zenml/cli/example.py [hunk 3]
```diff
@@ -147,7 +147,7 @@ def run_example(
                         "has no implementation for the "
                         "run method"
                     )
-                raise
+                # raise
```
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/cli/example.py:150]: WIP, just did that to see if this is the reason for my failing tests. This wasn't included before at all and lead to some tests not working correctly

File: tests/integration/test_examples_on_kubeflow.py (hunk 2 above)
Comment (schustmi) [reply tests/integration/test_examples_on_kubeflow.py]: I actually merged them into a single function and used `pytest.mark.parametrize` for now. That way it's easy to enable kubeflow but we can also quickly split it into two functions if necessary

File: src/zenml/cli/example.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/cli/example.py]: Reincluded the raising here

**Review by AlexejPenner — CHANGES_REQUESTED** [review]
Git shows that an empty file with name 'asd' was added, I guess that was accidental?

I like the code overall, I just added some tiny CR's regarding docstrings and typing on test functions.

**Comment by schustmi** [comment]
@AlexejPenner Maybe the `asd` file is similar to airflows `dag` comment 😉 
I removed it

**Review by AlexejPenner — APPROVED** [review]
Good to go from my side :)

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_component.py [hunk 4]
```diff
@@ -109,8 +108,23 @@ def __init__(
...
+        # This metadata file needs to be in a mounted emptyDir to avoid
+        # sporadic failures with the (not mature) PNS executor
+        # See these links for more information about limitations of PNS +
+        # security context:
+        # https://www.kubeflow.org/docs/components/pipelines/installation/localcluster-deployment/#deploying-kubeflow-pipelines
+        # https://argoproj.github.io/argo-workflows/empty-dir/
+        # KFP will switch to the Emissary executor (soon), when this emptyDir
+        # mount will not be necessary anymore, but for now it's still in alpha
+        # status (https://www.kubeflow.org/docs/components/pipelines/installation/choose-executor/#emissary-executor)
+        metadata_ui_path = "/outputs/mlpipeline-ui-metadata.json"
```
Comment (stefannica) [inline src/zenml/integrations/kubeflow/orchestrators/kubeflow_component.py:122]: I had to change this from `/mlpipeline-ui-metadata.json` to `/tmp/mlpipeline-ui-metadata.json` recently because the local user didn't have permissions to write to the root directory. Did you check that this file is actually created by the pipeline steps ? You should see something in the step visualizations in the Kubeflow UI, if it works.

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py [hunk 5]
```diff
@@ -56,11 +56,30 @@
     component_flavor=OrchestratorFlavor.KUBEFLOW,
 )
 class KubeflowOrchestrator(BaseOrchestrator):
-    """Orchestrator responsible for running pipelines using Kubeflow."""
+    """Orchestrator responsible for running pipelines using Kubeflow.
+
+    Attributes:
```
Comment (stefannica) [inline src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py:61]: Thank you for documenting these !

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py [hunk 6]
```diff
@@ -240,6 +259,11 @@ def _upload_and_run_pipeline(
                 logger.info(
                     "Started one-off pipeline run with ID '%s'.", result.run_id
                 )
+
+                if self.synchronous:
```
Comment (stefannica) [inline src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py:263]: I just thought I'd mention this: this is something I often needed even as an end user: forcing zenml to wait locally until the remote pipeline execution is complete. I understand this is only meant for integration testing, but it would be nice to have it exposed in the future as a feature that users can also access and use, maybe as a pipeline run config attribute.

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py [hunk 7]
```diff
@@ -291,7 +315,7 @@ def list_manual_setup_steps(
         """Logs manual steps needed to setup the Kubeflow local orchestrator."""
         global_config_dir_path = zenml.io.utils.get_global_config_directory()
         kubeflow_commands = [
-            f"> k3d cluster create CLUSTER_NAME --registry-create {container_registry_name} --registry-config {container_registry_path} --volume {global_config_dir_path}:{global_config_dir_path}\n",
+            f"> k3d cluster create CLUSTER_NAME --image {local_deployment_utils.K3S_IMAGE_NAME} --registry-create {container_registry_name} --registry-config {container_registry_path} --volume {global_config_dir_path}:{global_config_dir_path}\n",
```
Comment (stefannica) [inline src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py:318]: Just out of curiosity: why did you need to specify the image here explicitly ?

File: tests/integration/conftest.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+) -> Generator[Repository, None, None]:
+    """Creates a repo with a locally provisioned kubeflow stack.
+
+    As the resource provisioning for the local kubeflow deployment takes quite
+    a while, this fixture has a module scope and will therefore only run once.
+
+    **Note**: The fixture should not be used directly. Use the
+    `clean_kubeflow_repo` fixture instead that builds on top of this and
+    provides the  test with a clean working directory and artifact/metadata
+    store.
```
Comment (stefannica) [inline tests/integration/conftest.py:40]: Missing docstring for arguments and return value.

File: .github/workflows/kubeflow.yml [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+          python -m pip install --upgrade pip
+          python -m pip install "poetry"
+
+      - name: Install Dependencies
+        run: |
+          python -m poetry install
+          python -m poetry run zenml integration install kubeflow -f
+
+      - name: Build docker image
+        # this step can be much faster if we use an image that copies the
```
Comment (stefannica) [inline .github/workflows/kubeflow.yml:53]: Very important. Maybe you should even make this a TODO item.

File: src/zenml/integrations/kubeflow/docker_utils.py [hunk 10]
```diff
@@ -283,6 +286,20 @@ def get_image_digest(image_name: str) -> Optional[str]:
...
 
+def is_local_image(image_name: str) -> bool:
+    """Returns whether an image was pulled from a registry or not."""
+    docker_client = DockerClient.from_env()
+    images = docker_client.images.list(name=image_name)
+    if images:
+        # An image with this name is available locally -> now check whether it
+        # was pulled from a repo or built locally (in which case the repo
+        # digest is empty)
+        return get_image_digest(image_name) is None
```
Comment (stefannica) [inline src/zenml/integrations/kubeflow/docker_utils.py:297]: Very nice ! You have no idea how much this has been bugging me... or maybe you do :)

File: tests/integration/test_examples.py [hunk 11]
```diff
@@ -50,279 +50,117 @@ def example_runner(examples_dir):
...
+    ),
+    ExampleIntegrationTestConfiguration(
+        name="drift_detection",
+        validation_function=drift_detection_example_validation,
+    ),
+    ExampleIntegrationTestConfiguration(
+        name="mlflow_tracking",
+        validation_function=mlflow_tracking_example_validation,
+    ),
+    # TODO [HIGH]: Enable running the whylogs example on kubeflow
```
Comment (stefannica) [inline tests/integration/test_examples.py:117]: Is this because the whylogs example is currently failing on Kubeflow ?

File: tests/integration/test_examples.py [hunk 12]
```diff
@@ -50,279 +50,117 @@ def example_runner(examples_dir):
...
+    [pytest.param(example, id=example.name) for example in examples],
+)
+@pytest.mark.parametrize(
+    "repo_fixture_name", ["clean_repo"]  # , "clean_kubeflow_repo"]
+)
+def test_run_example(
+    example_configuration: ExampleIntegrationTestConfiguration,
+    repo_fixture_name: str,
+    request: pytest.FixtureRequest,
+) -> None:
```
Comment (stefannica) [inline tests/integration/test_examples.py:135]: I just love how you formalized everything into a single function. This is some next-level pytest magic.

File: tests/integration/test_examples.py [hunk 13]
```diff
@@ -50,279 +50,117 @@ def example_runner(examples_dir):
...
+    # ),
+]
+
+
+@pytest.mark.parametrize(
+    "example_configuration",
+    [pytest.param(example, id=example.name) for example in examples],
+)
+@pytest.mark.parametrize(
+    "repo_fixture_name", ["clean_repo"]  # , "clean_kubeflow_repo"]
```
Comment (stefannica) [inline tests/integration/test_examples.py:129]: left-over comment ?

File: tests/integration/conftest.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+from zenml.repository import Repository
+from zenml.stack import Stack
+
+
+@pytest.fixture(scope="module")
+def shared_kubeflow_repo(
+    base_repo: Repository,
+    tmp_path_factory: pytest.TempPathFactory,
+    module_mocker: MockerFixture,
+) -> Generator[Repository, None, None]:
```
Comment (stefannica) [inline tests/integration/conftest.py:31]: Is there a way we could keep the ability to run tests without Kubeflow ? Not all OS/archs support Kubeflow and even on those that do, I see some value for running these tests locally without involving kubeflow.

**Review by stefannica — CHANGES_REQUESTED** [review]
Great work formalizing everything around our integration testing to support kubeflow (and fixing some examples in the process too).  I had a couple of minor change requests, the other being mostly questions.

File: .github/workflows/kubeflow.yml [hunk 15]
```diff
@@ -0,0 +1,58 @@
+name: Kubeflow
+on: workflow_dispatch
```
[the remark this answers is not in the document]
Comment (schustmi) [reply .github/workflows/kubeflow.yml:2]: The kubeflow integration tests are disabled/not working at the moment due to some dependency conflicts until we merge Alexej's virtualenv PR. I kept this here as a (disabled) template for us to quickly activate them as soon as that's done

File: .gitignore [hunk 16]
```diff
@@ -176,6 +176,3 @@ zenml_examples/
 
 # Prettier
 .prettierignore
-
-# Docs
-docs/sphinx_docs/_build
```
[the remark this answers is not in the document]
Comment (schustmi) [reply .gitignore:181]: I think this was just leftover from back in the days when we used sphinx to generate our docs, now that we swapped to mkdocs it's not necessary anymore. Should have probably separated that into another PR as it's entirely unrelated

File: examples/run_example.sh [hunk 17]
```diff
@@ -57,6 +59,12 @@ main() {
         pre_run
       fi
     fi
+
+    if [ -n "$SETUP_STACK" ]; then
+      if [[ $(type -t setup_stack) == function ]]; then
+          setup_stack
+      fi
+    fi
```
[the remark this answers is not in the document]
Comment (schustmi) [reply examples/run_example.sh:67]: Very good point, we should probably (optionally?) deprovision the stack after the example is run. Do you think we can include that in the `cleanup` function that you implemented in your PR or should we create a separate one just for the stack?

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py (hunk 6 above)
Comment (schustmi) [reply src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py:263]: As you said it's not really intended for users right now (e.g. fixed timeout), but if you see a need for that it would certainly be an easy change. I'll add a todo for now if that's okay with you

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py (hunk 7 above)
Comment (schustmi) [reply src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py:318]: Every k3d release comes bundled with a default k3s image (which in turn refers to a kubernetes version). Users with a newer k3d installation had problems as the default image had a kubernetes version >= 1.22.0 and KFP is only compatible with < 1.22.0

File: .github/workflows/kubeflow.yml (hunk 9 above)
Comment (schustmi) [reply .github/workflows/kubeflow.yml:53]: Unfortunately our todo script is too stupid to detect non-python files atm :P
I'll create a jira issue directly to make sure we implement this soon

File: .github/workflows/kubeflow.yml (hunk 15 above)
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .github/workflows/kubeflow.yml:2]: Once this is merged I will incorporate it at the right place into my branch :)

File: .gitignore (hunk 16 above)
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .gitignore:181]: I must have missed this one when we transitioned to mkdocs

File: src/zenml/integrations/kubeflow/docker_utils.py (hunk 10 above)
Comment (schustmi) [reply src/zenml/integrations/kubeflow/docker_utils.py:297]: Trust me, I do .. :)

File: tests/integration/test_examples.py (hunk 11 above)
Comment (schustmi) [reply tests/integration/test_examples.py:117]: Exactly, but as long as we're not running on kubeflow I think this can be re-enabled

File: tests/integration/test_examples.py (hunk 13 above)
Comment (AlexejPenner) [reply tests/integration/test_examples.py:129]: I believe this one's for me to uncomment once its merged and we activate the tests

Comment (schustmi) [reply tests/integration/test_examples.py:129]: Left-over comment or hint on how to enable running the tests on kubeflow ;) I'll remove it

File: tests/integration/conftest.py (hunk 14 above)
Comment (schustmi) [reply tests/integration/conftest.py:31]: Yes I do have some ideas on how to make this distinction if we need it, e.g. pytest allows you to custom command line options (https://docs.pytest.org/en/6.2.x/example/simple.html#dynamically-adding-command-line-options). We could run our integration tests with `pytest tests/integration --with-kubeflow` or something similar to specify whether kubeflow should be included

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_component.py (hunk 4 above)
Comment (schustmi) [reply src/zenml/integrations/kubeflow/orchestrators/kubeflow_component.py:122]: Works for me, it shows the Markdown with all the execution parameters/inputs/outputs as usual

Comment (stefannica) [reply src/zenml/integrations/kubeflow/orchestrators/kubeflow_component.py:122]: Great !

File: src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py (hunk 6 above)
Comment (stefannica) [reply src/zenml/integrations/kubeflow/orchestrators/kubeflow_orchestrator.py:263]: a TODO was exactly what I was looking for

**Review by stefannica — APPROVED** [review]

File: examples/run_example.sh (hunk 17 above)
[the remark this answers is not in the document]
Comment (stefannica) [reply examples/run_example.sh:67]: The stack cleanup should be separate from the `post_run` function in my PR. It should be a `stack_cleanup` function, similar to this one, that is called at the end of the test run, after `post_run`.
